admin/services/prepare-check.service.js

```diff
@@ -0,0 +1,67 @@
+'use strict'
+
+const moment = require('moment')
+const R = require('ramda')
+const redisService = require('./redis-cache.service')
+
+const service = {
+  /**
+   * @param {[{object}]} checks the pupil checks to prepare
+   * @returns {Promise<void>}
+   */
+  prepareChecks: (checks) => {
+    if (!Array.isArray(checks)) {
+      throw new Error('checks is not an array')
+    }
+    const preparedChecks = checks.map(check => {
+      const preparedCheck = constructPreparedCheck(check)
+      return {
+        key: buildKey(check.schoolPin, check.pupilPin),
+        value: preparedCheck,
+        ttl: secondsBetweenNowAndPinExpiryTime(preparedCheck.pinExpiresAt)
+      }
+    })
+    return redisService.setMany(preparedChecks)
+  }
+}
+
+/**
+ *
+ * @param {Date} pinExpiry
+ * @returns {number} number of seconds between now and the pin expiry time
+ */
+function secondsBetweenNowAndPinExpiryTime (pinExpiry) {
+  const a = moment()
+  const b = moment(pinExpiry)
+  const differenceInSeconds = b.diff(a, 'seconds')
+  return differenceInSeconds
+}
+
+/**
+ * Return an entity suitable for inserting into the `preparedCheck` table
+ * @param {object} check
+ * @return {object}
+ */
+function constructPreparedCheck (check) {
+  const entity = {
+    checkCode: check.pupil.checkCode,
+    config: check.config,
+    createdAt: new Date(),
+    pinExpiresAt: moment(check.pupil.pinExpiresAt).toDate(),
+    pupil: R.omit(['id', 'checkFormAllocationId', 'pinExpiresAt'], check.pupil),
+    pupilId: check.pupil.id,
+    questions: check.questions,
+    school: check.school,
+    schoolId: check.school.id,
+    tokens: check.tokens,
+    updatedAt: new Date(),
+    pinValidFrom: new Date()
```

Comment:
Pins are only valid from some known time in the morning. 8AM?  This time should be local time for the school's timezone.

Reply:
sounds like new functionality is being introduced.
This will need to be raised with the PO

Reply:
This is what Paul was saying when we were discussing about MOD schools . He says even though we allowed to generate pin before 8 AM...but those pin should be valid from 8:00AM. As far as i know as soon as we generate pin, pupil can login from that point.

Reply:
fair enough

Reply:
I’ll remove this in light of what we discussed earlier.  It has its own ticket now

Reply:
`pinValidFrom` removed, and `updatedAt` and `createdAt` set to moments.